ONSdigital/census-contact-centre-service
# Cr 1080 Get /cases/{caseId} to return the latest of either the RM case or the cached case

# Motivation and Context

In the getCaseById endpoint it was previously only looking in the Firestore cache, to find a case by its id, if it had not already been found in the RM case service. However, it is now possible that there might exist a case, in Firestore, which is more up to date that the one returned by the case service. So the getCaseById endpoint needs to look in both places and return the case that it finds was most recently created or updated.

The CR-1080 ticket describes the task as follows:

The GET /cases/

{caseId}
endpoint currently fetches the case for Serco from RM case API, and only goes to the CC cache when RM does NOT return a case for us.

In the situation where Serco have modified a case using the PUT /cases endpoint, we will have the updated address etc in the cache, but currently we would not send that to Serco if RM return us the original case details.

Instead of 'RM and if not in RM use cache', additionally if RM DO have the case, we will ALSO look in the cache, and return whichever is the latest version as indicated by :

RM-CASE.lastUpdatedDateTime
CC-CASE:createdDateTime

If the CC receives multiple update request from Serco, because we store the cached case using the caseId as the primary key in firestore, we will only have one case in the cache - the last one updated.

# What has changed

The getCaseById endpoint has been modified so that it now works as described above. Four JUnit tests have also been written for the new functionality. These are named as follows:
1. testGetLatestFromCacheWhenResultsFromBothRmAndCache()
2. testGetLatestFromCacheWhenResultsFromBothRmAndCacheWithSmallTimeDifference()
3. testGetLatestFromRmWhenResultsFromBothRmAndCacheWithSmallTimeDifferences()
4. testGetLatestFromRmWhenResultsFromBothRmAndCache()

# How to test?


Run rabbitmq, the mock case service, and this branch of the contact centre service.

Use Postman to send a GET request to the following URL:

http://localhost:8171/cases/3305e937-6fb1-4ce1-9d4c-077f147789ac

Confirm that the address it returns is for Napier House, which means that it is the case from the Mock Case service

Now use the getCaseByUprn endpoint to create a case in Firestore by sending a GET request as follows:

http://localhost:8171/cases/uprn/100040239948

There should now exist, in Forestore, as case with the address "1 West Grove Road". Modify the id of the case (in Firestore) so that it is the same as the one for Napier House i.e. 3305e937-6fb1-4ce1-9d4c-077f147789ac

Now retry running the original request:

http://localhost:8171/cases/3305e937-6fb1-4ce1-9d4c-077f147789ac

Instead of Napier House the getCaseById endpoint should now return the more recently created Firestore case i.e. 1 West Grove Road
